cdap-explore/src/main/java/co/cask/cdap/explore/security/TokenSecureStoreUpdater.java

```diff
@@ -0,0 +1,161 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.explore.security;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.security.YarnTokenUtils;
+import co.cask.cdap.data.security.HBaseTokenUtils;
+import co.cask.cdap.explore.service.ExploreServiceUtils;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.security.User;
+import org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge20S;
+import org.apache.hadoop.mapreduce.MRConfig;
+import org.apache.hadoop.security.Credentials;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.yarn.conf.YarnConfiguration;
+import org.apache.twill.api.RunId;
+import org.apache.twill.api.SecureStore;
+import org.apache.twill.api.SecureStoreUpdater;
+import org.apache.twill.filesystem.LocationFactory;
+import org.apache.twill.internal.yarn.YarnUtils;
+import org.apache.twill.yarn.YarnSecureStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A {@link SecureStoreUpdater} that updates all secure tokens used by the platform.
+ */
+public final class TokenSecureStoreUpdater implements SecureStoreUpdater {
+  private static final Logger LOG = LoggerFactory.getLogger(TokenSecureStoreUpdater.class);
+  private final Configuration hConf;
+  private final YarnConfiguration yarnConf;
+  private final LocationFactory locationFactory;
+  private final long updateInterval;
+  private final boolean secureExplore;
+  private long nextUpdateTime = -1;
+  private Credentials credentials;
+
+  @Inject
+  public TokenSecureStoreUpdater(Configuration hConf, CConfiguration cConf, YarnConfiguration yarnConf,
+                                 LocationFactory locationFactory) {
+    this.hConf = hConf;
+    this.yarnConf = yarnConf;
+    this.locationFactory = locationFactory;
+    secureExplore = cConf.getBoolean(Constants.Explore.EXPLORE_ENABLED) && UserGroupInformation.isSecurityEnabled();
+    credentials = new Credentials();
+    updateInterval = calculateUpdateInterval();
+  }
+
+  private void refreshCredentials() {
+    try {
+      Credentials refreshedCredentials = new Credentials();
+
+      if (User.isHBaseSecurityEnabled(hConf)) {
+        HBaseTokenUtils.obtainToken(hConf, refreshedCredentials);
+      }
+
+      if (secureExplore) {
+        HiveTokenUtils.obtainToken(refreshedCredentials);
+        YarnTokenUtils.obtainToken(hConf, refreshedCredentials);
+        JobHistoryServerTokenUtils.obtainToken(hConf, refreshedCredentials);
+      }
+
+      YarnUtils.addDelegationTokens(hConf, locationFactory, refreshedCredentials);
```

Comment:
Add renewal time for HDFS token (`dfs.namenode.delegation.token.renew-interval`) too while calculating update interval below.